ir/v1.cpp

```diff
@@ -141,3 +142,29 @@ const IR::Type *IR::Primitive::inferOperandType(int operand) const {
             return IR::Type::Bits::get(32); }
     return IR::Type::Unknown::get();
 }
+
+Util::Enumerator<const IR::IDeclaration*>* IR::V1Program::getDeclarations() const {
+    using ScopeVal = decltype(scope)::value_type;
+
+    std::function<const IR::Node*(const ScopeVal&)> toNode =
+      [](const ScopeVal& value) -> const IR::Node* { return value.second; };
+    std::function<const IR::IDeclaration*(const IR::Node* const&)> toIDeclaration =
+      [](const IR::Node* const& node) { return node->to<IR::IDeclaration>(); };
+
+    // Return an enumerator over all Nodes in the scope which are IDeclarations.
+    return Util::Enumerator<ScopeVal>::createEnumerator(scope.begin(), scope.end())
+              ->map(toNode)
+              ->where([](const IR::Node* node) { return node->is<IR::IDeclaration>(); })
+              ->map(toIDeclaration);
+}
+
+const IR::IDeclaration* IR::V1Program::getDeclByName(cstring name) const {
```

Comment:
The other problem is that most of the P4_14 top-level named objects don't inherit from IDeclaration, though they could easily be modified to do so.  `HeaderOrMetadata`, `V1Parser`, `Attached`, `ActionFunction`, `V1Table`, and `V1Control` would all need it.